taar/recommenders/collaborative_recommender.py

```diff
@@ -72,14 +72,20 @@ def can_recommend(self, client_data, extra_data={}):
         return False
 
     def recommend(self, client_data, limit, extra_data={}):
+        """We only get meaningful recommendation if a client has at least an
+        addon installed."""
+        if len(client_data.get('installed_addons', [])) == 0:
+            return []
+
         # Addons identifiers are stored as positive hash values within the model.
-        installed_addons =\
+        installed_addons_as_hashes =\
             [positive_hash(addon_id) for addon_id in client_data.get('installed_addons', [])]
 
         # Build the query vector by setting the position of the queried addons to 1.0
         # and the other to 0.0.
-        query_vector = np.array([1.0 if (entry.get("id") in installed_addons) else 0.0
-                                 for entry in self.raw_item_matrix])
+        query_vector = np.array([1.0
```

Comment:
Agreed, this is more readable this way!

Reply:
The largest number of changes I've made are in the CollaborativeRecommender where the dictionary lookups didn't seem to match up to the production JSON data blobs in S3.  

w.r.t PEP8 patches - I'm just making those as I hit individual files.

Reply:
r? @Dexterp37